### PR TITLE
removing support in llvm 6 and below

### LINKED ISSUE #2560: Remove LLVM6 support from Glow
LLVM 7 is the latest LLVM version and LLVM 8 is not too far away. This we should drop any LLVM 6 support to simplify the code.

Specifically, we need to do the following:
- Adjust CMake files to use LLVM >=7 by default and bail if LLVM < 7 is used
- Remove any conditionally compiled LLVM 6-specific code inside #ifdefs

### PR DESCRIPTION
**Pull request by mortzur**
*Description*:
Removing code support in LLVM package versions <= 6.
*Edited*: CI scripts were modified to use LLVM version 8.
The reason for skipping LLVM-7 is a problematic llvm-7 distribution for Ubuntu-16.04 (xenial), which has been built with old C++ ABI and therefore is incompatible. Locally (MacOS) tests pass also on llvm-7.   

Should be followed up (in next commit) by removing all FACEBOOK_INTERNAL branches (with the specific patch number).

*Testing*:
* cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ../glow
   -- Found LLVM 7.0.1
    ...
   ninja test

* cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ../glow -DCMAKE_PREFIX_PATH=/usr/local/Cellar/llvm/8.0.0/
-- Found LLVM 8.0.0
   ...
   ninja test

* On a devserver (changes copied manually, llvm version 8):
   buck build //glow/...
   buck test //glow/...

*Documentation*:

Fixes #2560

### CODE REVIEW

**Comment by rdzhabarov** [comment]
Fixes https://github.com/pytorch/glow/issues/2560

**Comment by rdzhabarov** [comment]
You need to fix this https://github.com/pytorch/glow/blob/master/.circleci/build.sh#L45

Also, rebased on the latest master.

File: lib/LLVMIRCodeGen/DebugInfo.cpp [hunk 1]
```diff
@@ -160,18 +160,10 @@ LLVMIRGen::getOrCreateFunctionDebugInfo(llvm::Function *F, llvm::DIScope *scope,
     auto *DIFunctionTy = DIBuilder_->createSubroutineType(
         DIBuilder_->getOrCreateTypeArray(paramTys));
     // Create a debug information for the current function.
-#if LLVM_VERSION_MAJOR < 8 || FACEBOOK_INTERNAL
     DIFunction = DIBuilder_->createFunction(
         scope, F->getName(), "", file, lineNo, DIFunctionTy,
         false /* internal linkage */, true /* definition */, lineNo,
         llvm::DINode::FlagPrototyped, true /* isOptimized */);
-#else
```
Comment (opti-mix) [inline lib/LLVMIRCodeGen/DebugInfo.cpp]: Why did you remove this snippet? This was introduced for LLVM >=8, IIRC.

File: lib/LLVMIRCodeGen/GlowJIT.cpp [hunk 2]
```diff
@@ -72,36 +72,17 @@ class NotifyLoadedFunctor {
...
   void operator()(llvm::orc::VModuleKey key,
                   const llvm::object::ObjectFile &obj,
                   const llvm::RuntimeDyld::LoadedObjectInfo &objInfo) {
     auto &loadedObj = obj;
     // Inform the debugger about the loaded object file. This should allow for
     // more complete stack traces under debugger. And even it should even enable
     // the stepping functionality on platforms supporting it.
-#if LLVM_VERSION_MAJOR < 8 || FACEBOOK_INTERNAL
     dbgRegistrationListener_->NotifyObjectEmitted(loadedObj, objInfo);
-#else
```
Comment (opti-mix) [inline lib/LLVMIRCodeGen/GlowJIT.cpp]: Why did you remove this snippet? This was introduced for LLVM >=8, IIRC.

File: lib/LLVMIRCodeGen/GlowJIT.cpp [hunk 3]
```diff
@@ -150,24 +131,12 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
...
           },
           [](Error Err) { cantFail(std::move(Err), "lookupFlags failed"); })),
-#if LLVM_VERSION_MAJOR < 8 || FACEBOOK_INTERNAL
       objectLayer_(ES_,
                    [this](llvm::orc::VModuleKey) {
                      return RTDyldObjectLinkingLayer::Resources{
                          std::make_shared<SectionMemoryManager>(), resolver_};
                    },
                    NotifyLoadedFunctor(this)),
-#else
```
Comment (opti-mix) [inline lib/LLVMIRCodeGen/GlowJIT.cpp]: Why did you remove this snippet? This was introduced for LLVM >=8, IIRC.

File: lib/LLVMIRCodeGen/GlowJIT.cpp [hunk 4]
```diff
@@ -135,7 +116,7 @@ GlowJIT::GlowJIT(llvm::TargetMachine &TM)
                      return RTDyldObjectLinkingLayer::Resources{
                          std::make_shared<SectionMemoryManager>(), resolver_};
                    }),
-#elif LLVM_VERSION_MAJOR > 6
+#else // LLVM_VERSION_MAJOR > 6
```
Comment (jfix71) [inline lib/LLVMIRCodeGen/GlowJIT.cpp]: This comment seems incorrect

File: lib/LLVMIRCodeGen/LLVMIRGen.cpp [hunk 5]
```diff
@@ -300,13 +290,10 @@ void LLVMIRGen::performCodeGen() {
 #if FACEBOOK_INTERNAL && LLVM_VERSION_PATCH < 20181009
     getTargetMachine().addPassesToEmitFile(
         PM, asmStream, llvm::TargetMachine::CodeGenFileType::CGFT_AssemblyFile);
-#elif LLVM_VERSION_MAJOR > 6
+#else // LLVM_VERSION_MAJOR > 6
```
Comment (jfix71) [inline lib/LLVMIRCodeGen/LLVMIRGen.cpp]: This comment seems incorrect.

File: include/glow/LLVMIRCodeGen/GlowJIT.h [hunk 6]
```diff
@@ -49,18 +49,13 @@ class GlowJIT {
   SymbolStringPool SSP_;
   ExecutionSession ES_;
   std::shared_ptr<SymbolResolver> resolver_;
-#elif LLVM_VERSION_MAJOR > 6
+#else // LLVM_VERSION_MAJOR > 6
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/GlowJIT.h]: This comment seems incorrect.

File: lib/LLVMIRCodeGen/DebugInfo.cpp (hunk 1 above)
Comment (mortzur) [reply lib/LLVMIRCodeGen/DebugInfo.cpp]: This passed in the internal version, therefore I wrongfully assumed it passes for llvm8. It doesn't built locally. reverting all of these patches.

File: lib/LLVMIRCodeGen/GlowJIT.cpp (hunk 2 above)
Comment (mortzur) [reply lib/LLVMIRCodeGen/GlowJIT.cpp]: This passed in the internal version, therefore I wrongfully assumed it passes for llvm8. It doesn't built locally. reverting all of these patches.

File: lib/LLVMIRCodeGen/GlowJIT.cpp (hunk 3 above)
Comment (mortzur) [reply lib/LLVMIRCodeGen/GlowJIT.cpp]: This passed in the internal version, therefore I wrongfully assumed it passes for llvm8. It doesn't built locally. reverting all of these patches.

File: lib/LLVMIRCodeGen/GlowJIT.cpp (hunk 4 above)
Comment (mortzur) [reply lib/LLVMIRCodeGen/GlowJIT.cpp]: Hmm. The comment refers to versions for which the following section should work.
Do you mean that I'm not actually enforcing the condition anywhere?

File: lib/LLVMIRCodeGen/DebugInfo.cpp [hunk 7]
```diff
@@ -160,18 +160,19 @@ LLVMIRGen::getOrCreateFunctionDebugInfo(llvm::Function *F, llvm::DIScope *scope,
     auto *DIFunctionTy = DIBuilder_->createSubroutineType(
         DIBuilder_->getOrCreateTypeArray(paramTys));
     // Create a debug information for the current function.
-#if LLVM_VERSION_MAJOR < 8 || FACEBOOK_INTERNAL
+#if LLVM_VERSION_MAJOR == 7
```
Comment (opti-mix) [inline lib/LLVMIRCodeGen/DebugInfo.cpp]: Is it OK that FACEBOOK_INTERNAL was removed?

**Comment by opti-mix** [comment]
> On a devserver (changes copied manually, llvm version 8):
buck build //glow/...
buck test //glow/...

Does it also build with the currently available LLVM on devservers? I think the default LLVM on devservers is not 8.0 yet, or?

**Comment by mortzur** [comment]
> > On a devserver (changes copied manually, llvm version 8):
> > buck build //glow/...
> > buck test //glow/...
> 
> Does it also build with the currently available LLVM on devservers? I think the default LLVM on devservers is not 8.0 yet, or?

yes and yes.

File: lib/LLVMIRCodeGen/BundleSaver.cpp [hunk 8]
```diff
@@ -178,13 +174,10 @@ void BundleSaver::produceBundle(llvm::StringRef outputDir) {
 #if FACEBOOK_INTERNAL && LLVM_VERSION_PATCH < 20181009
     TM.addPassesToEmitFile(
         PM, outputFile, llvm::TargetMachine::CodeGenFileType::CGFT_ObjectFile);
-#elif LLVM_VERSION_MAJOR > 6
+#else // LLVM_VERSION_MAJOR > 6
```
Comment (rdzhabarov) [inline lib/LLVMIRCodeGen/BundleSaver.cpp]: why `#else` assumes it's > 6. I do not think this is correct.

File: lib/LLVMIRCodeGen/GlowJIT.cpp (hunk 4 above)
Comment (rdzhabarov) [reply lib/LLVMIRCodeGen/GlowJIT.cpp]: yes the condition is not enforced anywhere.

Comment (jfix71) [reply lib/LLVMIRCodeGen/GlowJIT.cpp]: I was referring to the other side of this branch not having anything to do with `LLVM_VERSION_MAJOR > 6`, i.e. it's `#if FACEBOOK_INTERNAL && LLVM_VERSION_PATCH < 20181009` vs. this `#else`

File: .circleci/build.sh [hunk 9]
```diff
@@ -24,30 +24,18 @@ install_pocl() {
...
-    export PATH="$SCCACHE_BIN_DIR:$PATH"
-fi
-
-GLOW_DIR=$PWD
-
 # Install Glow dependencies
+sudo apt-add-repository "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-8 main"
 sudo apt-get update
-sudo apt-get install -y llvm-6.0 llvm-6.0-dev libpng-dev libgoogle-glog-dev
+sudo apt-get install -y llvm-8 clang-8 llvm-8-dev libclang-8-dev libpng-dev libgoogle-glog-dev opencl-headers
```
Comment (rdzhabarov) [inline .circleci/build.sh]: nit: why do we need to install opencl-headers for non open cl build?

File: lib/Backends/CPU/CMakeLists.txt [hunk 10]
```diff
@@ -4,11 +4,17 @@ else()
...
-               llvm-link)
+if(TARGET llvm-link)
+  set(LLVM_LINK_BIN $<TARGET_FILE:llvm-link>)
+else()
+  find_program(LLVM_LINK_BIN
+               NAMES
+                 llvm-link-${LLVM_VERSION_MAJOR}
+                 llvm-link-8
+                 llvm-link-7
+                 llvm-link-6.0
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CMakeLists.txt]: should we get rid of 6.0 here?

File: lib/LLVMIRCodeGen/DebugInfo.cpp (hunk 7 above)
Comment (mortzur) [reply lib/LLVMIRCodeGen/DebugInfo.cpp]: fixing

File: tests/unittests/QuantizationTest.cpp [hunk 11]
```diff
@@ -117,11 +117,13 @@ TEST(Quantization, Serialize) {
   testSerialization(expected);
 }
 
+#if LLVM_VERSION_MAJOR < 8
```
Comment (nadavrot) [inline tests/unittests/QuantizationTest.cpp]: Is this a use case that we really want to test? If it doesn't work on LLVM8 then maybe we don't really care about this use case and can simply delete it?

Comment (mortzur) [reply tests/unittests/QuantizationTest.cpp]: I'm just not sure if this is a bug or a valid behavior: https://github.com/pytorch/glow/issues/2588

File: .circleci/build.sh (hunk 9 above)
Comment (mortzur) [reply .circleci/build.sh]: The DEBUG case uses openCL , installs pocl which needs these installations .
Do you think it would be better to install only for DEBUG?
Do we plan to use openCL also for non-debug cases?

Comment (rdzhabarov) [reply .circleci/build.sh]: I think it's cleaner just add this into `install_pocl` function.

Comment (rdzhabarov) [reply .circleci/build.sh]: > Do we plan to use openCL also for non-debug cases?

We could, but no matter what all openCL builds will invoke `install_pocl`.

Comment (mortzur) [reply .circleci/build.sh]: Fixed

File: tests/unittests/QuantizationTest.cpp (hunk 11 above)
Comment (nadavrot) [reply tests/unittests/QuantizationTest.cpp]: Okay.

**Comment by rdzhabarov** [comment]
Have you tried running these changes on fbcode?

**Comment by rdzhabarov** [comment]
One of the tests failed: 25/34 Test #28: QuantizationTest ....................Child aborted***Exception:  22.27 sec

You could check the master branch results. Can you check what's going on?
